### ISSUE TITLE
consumer.emit('done') called once for each broker?

### ISSUE THREAD

**Issue by pcothenet** [SOHU-Co/kafka-node#440]
I'm still trying to build a batch consumer that fetches a bunch of message, process them asynchronously and commit only once the message have been processed.

A seemingly ideal solution would be #371 but in the meantime, here's what I'm doing:

1. on the first message, pause the consumer
2. accumulate the records
3. once all records fetched have been accumulated, process them, commit the offsets, empty the accumulator and resume the consumer

To know that all messages from the fetch have been emitted, I rely on the (undocumented) 'done' event on the consumer. However, it seems it's emitted several time per fetch. I looked at the code and it seems that for each fetchRequest, the 'done' event is emitted once per broker. Can you please confirm? 

(also, is that the expected behavior for that event?)

**Comment by hyperlink** [comment]
Hello @pcothenet. The client should only connect to [one broker](https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L89). And you should only see a `done` when all the messages have been emitted.

Looks like you're on the right track. Here's what our wrapper around the consumer looks like:

1. Create a `async.queue` with message processor and concurrency of one (the message processor itself is wrapped with `setImmediate` so it will not freeze up the event loop)
2. Set the `queue.drain` to resume the consumer
3. The handler for consumer's `message` event pauses the consumer and pushes the message to the queue.

And we let the consumer's own `autoCommit` to take care of the committing the offsets for us. This wrapper works well for us.

**Comment by pcothenet** [comment]
Hum, interesting. My kafka has three brokers.

Interesting. I ran a test with a single consumer and:

- `sendToBroker` [here](https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L420) is called three times for each `send`
- so the callback of the `decoder` gets called three times 
- so `done` was called three time for a single `sendFetchRequest`

Do you know what could cause this?

**Comment by pcothenet** [comment]
Thanks for sharing your method, I was able to do something similar on my end.

That said, is one `done` by broker the behavior you expect?

**Comment by zoluzo** [comment]
@pcothenet I got the same result. Consumer sends to all kafka brokers to fetch data.

**Comment by zoluzo** [comment]
@hyperlink  A consumer will fetch data from more than one kafka brokers if partitions are on different brokers.
I think it is not right that consumer.emit('done') called once for each broker

**Comment by gbarton** [comment]
Does anyone have an example gist of this method of getting effectively synchronous in order processing of messages via the async.queue style?  I've tried a few times and I cannot get exactly once processing yet.

**Comment by gbarton** [comment]
Alright, this might be working right, sanity check for those that know this better?  By wrapping the commit in a setImmediate, I appear to have synced commits now, is that because the callback is effectively called and trapped in the same tick as the message handling now?

```javascript
var count = 0;

var q = async.queue(function(message, callback) {
  // handle message
  debug('message count/offset', count, message.offset);
  count++;
  async.setImmediate(function() {
    consumer.commit(true, function(err, cResp) {
       return callback(err);
    }
  });
,1});

q.drain = function() {
 consumer.resume();
}

var consumer = new kafka.Consumer(client, topics, {groupId: 'foo', autoCommit: false, fromOffset:false});

consumer.on('message', function(message) {
  async.setImmediate(function() {
    consumer.pause();
    q.push(message, function(err) {
      // things dont work without a callback?
     debug('message handled');
    });
  });
});
```

**Comment by aikar** [comment]
After much research and fixing into #1072, this isn't worth worrying about or changing.

With newer kafka versions, we use a long polling based approach for fetching.
If you have 5 brokers, you will have 5 fetches pending all at same time.

When data comes in on 1 topic/partition, which ever broke is the leader for that, is which pending long poller will return to the consumer and process the data.

However, your other 4 are still pending, waiting for data.
The consumer recalls .fetch(), your broker that just returned data then reconnects, and the other 4 are skipped because they are already waiting.

This is desired, and not worth changing.

In my PR linked, I do add a new event, fetchprocessing, that I maintain a reference counter of active fetch processing to know to not start another one.

I believe my PR will resolve your concerns, or that this issue was no longer needed in the first place, and can be closed.
